CONTRIBUTING.md

```diff
@@ -160,6 +160,60 @@ You can also perform the same steps via the [GitHub CLI tool](https://cli.github
 Execute `gh pr create`, and step through the dialog to create the pull request.
 If the branch you will submit against is not the default branch, use the `-B {branch}` option to specify the branch to create the patch against.
 
+#### Performing assertions on the code
+
+We distinguish assertions in two categories:
+
+# Expectations of something that is always supposed to be true
+# Expectations of some input to follow some rule
+
+The former is a development-only concern, aiming to aid tools (e.g. IDEs, static analysis) to perform a more accurate type detection.
+In this scenario, you should the [`assert()`](https://www.php.net/manual/en/function.assert.php) function, which executes the expression when [`zend.assertions`](https://www.php.net/manual/en/ini.core.php#ini.zend.assertions) is enabled.
+That will make sure that the condition is valid, while easing future refactoring.
+
+Example:
+
+```php
+<?php
+declare(strict_types=true);
+
+namespace Laminas\Examples;
+
+use Psr\Container\ContainerInterface;
+
+use function assert;
+
+$container = require __DIR__ . '/../config/di-container.php';
+// we know that the file above ALWAYS returns an instance of this type
+assert($container instanceof ContainerInterface);
+
+// now we can call `ContainerInterface#get()`, which will be understood by psalm and even autocompleted by PHPStorm!
+$app = $container->get(MyApp::class);
```

Comment:
Maybe to make things clear, we should use `Assert::isInstanceOf($app, MyApp::class);` as thats exactly what @Ocramius mentioned during the meeting.

Trusting the Container to return the class which is requested is not possible due to configuration changes by upstream projects (if I understood correct).

So with this, we would have both use-cases in one example.
```suggestion
$app = $container->get(MyApp::class);
Assert::isInstanceOf($app, MyApp::class);
```

Reply:
I disagree here but probably the example isn't the best. I didn't want to use an example which is only relevant to psalm, can you think of something else for the `assert()` case?

Reply:
Huh? Maybe I don't understand the intention but I am fine with keeping the `assert` you added there.
Just thought that when adding `Assert` after fetching something from the container, we do have both examples on why we are differentiate between `Assert` and `assert` as `ContainerInterface#get` states that it returns `mixed`.

Reply:
@boesing sorry... I completely misread your comment and said that thing about psalm :facepalm: 

I usually have other means (tests) to guarantee the configuration of the container, which is why I often see that guard as a dev-only concern. However, that is indeed not always true.

I'll adapt things and push in a sec.